prometheus/go_collector.go

```diff
@@ -9,27 +10,245 @@ import (
 type goCollector struct {
 	goroutines Gauge
 	gcDesc     *Desc
+
+	// memstats object to reuse
+	ms *runtime.MemStats
+	// metrics to describe and collect
+	metrics memStatsMetrics
 }
 
 // NewGoCollector returns a collector which exports metrics about the current
 // go process.
 func NewGoCollector() *goCollector {
 	return &goCollector{
+		ms: new(runtime.MemStats),
 		goroutines: NewGauge(GaugeOpts{
-			Name: "go_goroutines",
-			Help: "Number of goroutines that currently exist.",
+			Namespace: "go",
+			Name:      "goroutines",
+			Help:      "Number of goroutines that currently exist.",
 		}),
 		gcDesc: NewDesc(
 			"go_gc_duration_seconds",
 			"A summary of the GC invocation durations.",
 			nil, nil),
+		metrics: memStatsMetrics{
+			{
+				desc: NewDesc(
+					memstatNamespace("alloc_bytes"),
+					"Number of bytes allocated and still in use.",
+					nil, nil,
+				),
+				eval:    func(ms *runtime.MemStats) float64 { return float64(ms.Alloc) },
+				valType: GaugeValue,
+			}, {
+				desc: NewDesc(
+					memstatNamespace("alloc_bytes_total"),
+					"Total number of bytes allocated, even if freed.",
+					nil, nil,
+				),
+				eval:    func(ms *runtime.MemStats) float64 { return float64(ms.TotalAlloc) },
+				valType: CounterValue,
+			}, {
+				desc: NewDesc(
+					memstatNamespace("sys_bytes"),
+					"Number of bytes obtained from system",
+					nil, nil,
+				),
+				eval:    func(ms *runtime.MemStats) float64 { return float64(ms.Sys) },
+				valType: CounterValue,
```

Comment:
I'm not sure. I will make it a GaugeValue for now. How would you go about figuring this out?